bin/validate_podspec

```diff
@@ -29,4 +60,4 @@ echo "--- :microscope: Validate Podspec"
 # https://github.com/Automattic/buildkite-ci/issues/7
 xcrun simctl list >> /dev/null
 
-bundle exec pod lib lint --verbose --fail-fast
+bundle exec pod lib lint "${COCOAPODS_FLAGS[@]}" -- "$@"
```

Comment:
What is the use of the extra `--` before the last parameter?

Reply:
Security, by separating flags from arguments and avoid allowing any user-provided parameter ("$@") to be interpreted as a flag… and allowing injection.

The idea was to only allow a specific and explicit subset of flags of the `pod lib lint` command (for example `--allow-warnings` or `--include-podspecs`) but not blindly permit any other random flag (like `--skip-tests` or `--allow-root`).

Otherwise I could have just used `bundle exec pod lib lint "$@"` directly on this line (and not parse explicit flags in `COCOAPODS_FLAGS` in the switch case at the start), and as a result, when calling `validate_podspec --allow-warnings MyPod.podspec` the `"$@"` would have been expanded by the shell to `"--allow-warnings" "MyPod.podspec`—and that would have worked the same as the current code that parses `--allow-warnings` into `COCOAPODS_FLAGS` and parses the rest (`MyPod.podspec`) in `$@` without needing extra code… BUT that would **also** have allowed anyone calling `validate_podspec --allow-root MyPod.podspec` and thus call `pod lib lint --allow-root MyPod.podpsec`, while I didn't want that option to be allowed.

With the `--`, if one tries to call `validate_podspec --allow-warnings --allow-root MyPod.podspec`, this will result in `bundle exec pod lib lint "--allow-warnings" -- "--allow-root" "MyPod.podspec"` once this line is expanded—with `COCOAPODS_FLAGS` containing `-allow-warnings` and `$@` containing `--allow-root MyPod.podspec". But thanks to the `--` I inserted in between, since `--allow-root` appears after that `--`, it will be interpreted as an argument—instead of risking of being interpreted as a flag because it starts with `--…`— and `pod lib lint` will interpret it as "I want to lint the pods named `--allow-root` and `MyPod.podspec`"—and will then fail because there's no file named `--allow-root`, successfully preventing using a flag we don't want to allo—instead of interpret it as "I want to lint the pod named `MyPod.podspec` but with the `--allow-root` option enabled"—and thus succeed… but also be open to potential root injections.

Reply:
See https://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html

> The argument -- terminates all options; any following arguments are treated as non-option arguments, even if they begin with a hyphen.

Reply:
> The idea was to only allow a specific and explicit subset of flags

I got that from the code, though I didn't realize the use of `--`. And actually... what would be the use case of other parameters at the end on `"$@"`?

> See https://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html
> 
> > The argument -- terminates all options; any following arguments are treated as non-option arguments, even if they begin with a hyphen.

TIL! Makes sense, thanks. That makes me wonder about other cases we have in `a8c-ci-toolkit-buildkite-plugin` with that potential breach that may be good to revisit.

Reply:
> And actually... what would be the use case of other parameters at the end on "$@"?

The name of the podspec to validate, if one wants to validate each podspec individually instead of keeping the default of `pod lib lint` linting every `*.podspec` it finds. This could potentially be useful if one wants to call `validate_podspec --allow-warnings A.podspec` but `validate_podspec B.podspec` without allowing warnings for `B`, for example.